gpgagent/gpgagent.go

```diff
@@ -0,0 +1,179 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Forked from https://camlistore.googlesource.com/camlistore/+/master/pkg/misc/gpgagent/
+
+// Package gpgagent interacts with the local GPG Agent.
+package gpgagent
+
+import (
+	"bufio"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"io"
+	"net"
+	"net/url"
+	"os"
+	"strings"
+)
+
+// Conn is a connection to the GPG agent.
+type Conn struct {
+	c  io.ReadWriteCloser
+	br *bufio.Reader
+}
+
+var (
+	ErrNoAgent = errors.New("GPG_AGENT_INFO not set in environment")
+	ErrNoData  = errors.New("GPG_ERR_NO_DATA cache miss")
+	ErrCancel  = errors.New("gpgagent: Cancel")
+)
+
+// NewConn connects to the GPG Agent as described in the
+// GPG_AGENT_INFO environment variable.
+func NewConn() (*Conn, error) {
+	sp := strings.SplitN(os.Getenv("GPG_AGENT_INFO"), ":", 3)
+	if len(sp) == 0 || len(sp[0]) == 0 {
+		return nil, ErrNoAgent
+	}
+	addr := &net.UnixAddr{Net: "unix", Name: sp[0]}
+	uc, err := net.DialUnix("unix", nil, addr)
+	if err != nil {
+		return nil, err
+	}
+	br := bufio.NewReader(uc)
+	lineb, err := br.ReadSlice('\n')
+	if err != nil {
+		return nil, err
+	}
+	line := string(lineb)
+	if !strings.HasPrefix(line, "OK") {
+		return nil, fmt.Errorf("gpgagent: didn't get OK; got %q", line)
+	}
+	return &Conn{uc, br}, nil
+}
+
+func (c *Conn) Close() error {
+	c.br = nil
+	return c.c.Close()
+}
+
+// PassphraseRequest is a request to get a passphrase from the GPG
+// Agent.
+type PassphraseRequest struct {
+	CacheKey, Error, Prompt, Desc string
+
+	// If the option --no-ask is used and the passphrase is not in
+	// the cache the user will not be asked to enter a passphrase
+	// but the error code GPG_ERR_NO_DATA is returned.  (ErrNoData)
+	NoAsk bool
+}
+
+func (c *Conn) RemoveFromCache(cacheKey string) error {
+	_, err := fmt.Fprintf(c.c, "CLEAR_PASSPHRASE %s\n", url.QueryEscape(cacheKey))
+	if err != nil {
+		return err
+	}
+	lineb, err := c.br.ReadSlice('\n')
+	if err != nil {
+		return err
+	}
+	line := string(lineb)
+	if !strings.HasPrefix(line, "OK") {
+		return fmt.Errorf("gpgagent: CLEAR_PASSPHRASE returned %q", line)
+	}
+	return nil
+}
+
+func (c *Conn) GetPassphrase(pr *PassphraseRequest) (passphrase string, outerr error) {
+	defer func() {
+		if e, ok := recover().(string); ok {
+			passphrase = ""
+			outerr = errors.New(e)
+		}
+	}()
+	set := func(cmd string, val string) {
+		if val == "" {
+			return
+		}
+		_, err := fmt.Fprintf(c.c, "%s %s\n", cmd, val)
+		if err != nil {
+			panic("gpgagent: failed to send " + cmd)
+		}
+		line, _, err := c.br.ReadLine()
+		if err != nil {
+			panic("gpgagent: failed to read " + cmd)
+		}
+		if !strings.HasPrefix(string(line), "OK") {
+			panic("gpgagent: response to " + cmd + " was " + string(line))
+		}
+	}
```

Comment:
What is the purpose of `panic` here? It looks like those could just be changed to `return` and the `recover` `defer` could be removed.